### PR TITLE
Add bifrost_append_and_seal_concurrent test

### PR DESCRIPTION
**Pull request by jackkleeman**

### CODE REVIEW

File: server/tests/replicated_loglet.rs [hunk 1]
```diff
@@ -175,4 +182,168 @@ mod tests {
         )
         .await
     }
+
+    #[test(tokio::test)]
```
Comment (AhmedSoliman) [inline server/tests/replicated_loglet.rs:186]: We probably want this to run in a multi_thread runtime for true concurrency

File: server/tests/replicated_loglet.rs [hunk 2]
```diff
@@ -175,4 +182,168 @@ mod tests {
...
+                    let (bifrost, metadata_writer, metadata_store_client) = (
+                        test_env.bifrost.clone(),
+                        test_env.metadata_writer.clone(),
+                        test_env.metadata_store_client.clone()
+                    );
+
+                    async move {
+                        let cancellation_token = cancellation_token();
+
+                        let mut chain = metadata.updateable_logs_metadata().map(|logs| logs.chain(&log_id).expect("a chain to exist"));
```
Comment (AhmedSoliman) [inline server/tests/replicated_loglet.rs:243]: metadata.logs_ref()).map... is probably what you actually need here. This gets you a wrapped reference of the last logs metadata.

File: server/tests/replicated_loglet.rs [hunk 3]
```diff
@@ -175,4 +182,168 @@ mod tests {
         )
         .await
     }
+
+    #[test(tokio::test)]
+    async fn bifrost_append_and_seal_concurrent() -> googletest::Result<()> {
+        const TEST_DURATION: Duration = Duration::from_secs(60);
```
Comment (AhmedSoliman) [inline server/tests/replicated_loglet.rs:188]: Wouldn't that have a very noticeable impact on the duration of our tests? Perhaps we can make this as 10 seconds by default and we seal every second?

File: server/tests/replicated_loglet.rs [hunk 4]
```diff
@@ -175,4 +182,168 @@ mod tests {
...
+                            while !cancellation_token.is_cancelled() {
+                                tokio::time::sleep(std::time::Duration::from_secs(10)).await;
+
+                                let mut params = ReplicatedLogletParams::deserialize_from(
+                                    chain.live_load().tail().config.params.as_ref(),
+                                )?;
+                                if last_loglet_id == Some(params.loglet_id) {
+                                    eprintln!("Skipping seal as metadata has not caught up from the last seal (version={})", metadata.logs_version());
+                                    continue;
+                                }
```
Comment (AhmedSoliman) [inline server/tests/replicated_loglet.rs:263]: Interesting! do you actually see this happening? IIRC, seal_and_extend_chain will write the updated metadata to our metadata manager before it returns, so we shouldn't really see this happening.

Comment (jackkleeman) [reply server/tests/replicated_loglet.rs:263]: no, i think this only happened when operator error meant that the seal didnt finish (eg when i didnt increment the segment index). what ill do instead is fail here

File: server/tests/replicated_loglet.rs (hunk 2 above)
Comment (jackkleeman) [reply server/tests/replicated_loglet.rs:243]: `metadata.logs_ref()).map` isnt a function - isnt logs_ref pinned, ie its a frozen arc from a point of time?

Comment (AhmedSoliman) [reply server/tests/replicated_loglet.rs:244]: it returns a reference of the last known value, and that's what you actually need here.

**Review by AhmedSoliman — APPROVED** [review]

File: server/tests/replicated_loglet.rs (hunk 2 above)
Comment (jackkleeman) [reply server/tests/replicated_loglet.rs:244]: hmm, but this is used in a loop, and i need the loglet id in params to change with each seal

Comment (AhmedSoliman) [reply server/tests/replicated_loglet.rs:244]: You are right, let's keep is as updateable in this case.

**Pull request merged**